packages/ember/tests/routing/substates_test.js

```diff
@@ -1,1042 +1,880 @@
+    assert.throws(() => {
+      this.visit('/grandma/mom/sally');
+    }, (err) => err.msg === 'did it broke?', 'it broke');
+  }
 
-  Router.map(function() {
-    this.route('foo', function() {
-      this.route('bar', { path: '/bar', resetNamespace: true }, function() {
-      });
-    });
-  });
+  ['@test Default error events move into nested route, prioritizing more specifically named error routes - NEW'](assert) {
+    this.addTemplate('grandma.error', 'ERROR: {{model.msg}}');
+    this.addTemplate('mom_error', 'MOM ERROR: {{model.msg}}');
+
+    this.add('route:mom.sally', Route.extend({
+      model() {
+        return RSVP.reject({
+          msg: 'did it broke?'
+        });
+      },
+      actions: {
+        error() {
+          return true;
+        }
+      }
+    }));
 
-  App.ApplicationController = Controller.extend();
+    assert.throws(() => {
+      this.visit('/grandma/mom/sally');
+    }, (err) => err.msg === 'did it broke?', 'it broke');
 
-  let deferred = RSVP.defer();
-  App.BarRoute = Route.extend({
-    model() {
-      return deferred.promise;
-    }
-  });
+    assert.equal(this.$('#app').text(), 'GRANDMA MOM ERROR: did it broke?',
+      'the more specifically named mome error substate was entered over the other error route'
+    );
 
-  bootApplication('/foo/bar');
+    let appController = this.getController('application');
 
-  equal(jQuery('#app', '#qunit-fixture').text(), 'BAR LOADING', 'foo.bar_loading was entered (as opposed to something like foo/foo/bar_loading)');
+    assert.equal(appController.get('currentPath'), 'grandma.mom_error',
+      'Initial route fully loaded'
+    );
+  }
 
-  run(deferred, 'resolve');
+  ['@test Slow promises waterfall on startup'](assert) {
+    let grandmaDeferred = RSVP.defer();
+    let sallyDeferred = RSVP.defer();
 
-  equal(jQuery('#app', '#qunit-fixture').text(), 'YAY');
-});
+    this.addTemplate('loading', 'LOADING');
+    this.addTemplate('mom', 'MOM {{outlet}}');
+    this.addTemplate('mom.loading', 'MOMLOADING');
+    this.addTemplate('mom.sally', 'SALLY');
 
-QUnit.test('Prioritized loading substate entry works with preserved-namespace nested routes', function() {
-  expect(2);
+    this.add('route:grandma', Route.extend({
+      model() {
+        return grandmaDeferred.promise;
+      }
+    }));
 
-  templates['foo/bar_loading'] = 'FOOBAR LOADING';
-  templates['foo/bar'] = 'YAY';
+    this.add('route:mom', Route.extend({
+      model() {
+        return {};
+      }
+    }));
 
-  Router.map(function() {
-    this.route('foo', function() {
-      this.route('bar');
-    });
-  });
+    this.add('route:mom.sally', Route.extend({
+      model() {
+        return sallyDeferred.promise;
+      },
 
-  App.ApplicationController = Controller.extend();
+      setupController() {
+      }
+    }));
 
-  let deferred = RSVP.defer();
-  App.FooBarRoute = Route.extend({
-    model() {
-      return deferred.promise;
-    }
-  });
+    this.visit('/grandma/mom/sally');
```

Comment:
Yeah that would works well. The thing I was reaching for was to chain the promises, and your example does that. Another alternative:

```js
let promise = this.visit('/grandma/mom/sally').then(() => {
  let text = this.$('#app').text();
});
model.resolve();
return promise;
```